File: app/src/main/java/com/nshell/nsplayer/ui/VideoListAdapter.java
package com.nshell.nsplayer.ui;

import android.graphics.Color;
import android.net.Uri;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.nshell.nsplayer.R;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.Set;

public class VideoListAdapter extends RecyclerView.Adapter<VideoListAdapter.ViewHolder> {
    public interface OnItemClickListener {
        void onItemClick(DisplayItem item);
    }

    public interface OnItemOverflowClickListener {
        void onOverflowClick(DisplayItem item);
    }

    public interface OnSelectionChangedListener {
        void onSelectionChanged(int selectedCount, boolean selectionMode);
    }

    private static final int VIEW_TYPE_DEFAULT = 0;
    private static final int VIEW_TYPE_VIDEO_LIST = 1;
    private static final int VIEW_TYPE_VIDEO_TILE = 2;
    private static final int VIEW_TYPE_TILE_DEFAULT = 3;

    private final List<DisplayItem> items = new ArrayList<>();
    private final Set<Integer> selectedPositions = new HashSet<>();
    private VideoDisplayMode videoDisplayMode = VideoDisplayMode.LIST;
    private OnItemClickListener clickListener;
    private OnItemOverflowClickListener overflowClickListener;
    private OnSelectionChangedListener selectionChangedListener;
    private boolean selectionMode = false;

    public void submit(List<DisplayItem> nextItems) {
        items.clear();
        if (nextItems != null) {
            items.addAll(nextItems);
        }
        clearSelectionInternal();
        notifyDataSetChanged();
    }

    public void setVideoDisplayMode(VideoDisplayMode mode) {
        if (mode != null) {
            this.videoDisplayMode = mode;
        }
    }

    public void setOnItemClickListener(OnItemClickListener listener) {
        this.clickListener = listener;
    }

    public void setOnItemOverflowClickListener(OnItemOverflowClickListener listener) {
        this.overflowClickListener = listener;
    }

    public void setOnSelectionChangedListener(OnSelectionChangedListener listener) {
        this.selectionChangedListener = listener;
    }

    public boolean isSelectionMode() {
        return selectionMode;
    }

    public int getSelectedCount() {
        return selectedPositions.size();
    }

    public boolean isAllSelected() {
        return !items.isEmpty() && selectedPositions.size() == items.size();
    }

    public void selectAll() {
        selectionMode = true;
        selectedPositions.clear();
        for (int i = 0; i < items.size(); i++) {
            selectedPositions.add(i);
        }
        notifySelectionChanged();
        notifyDataSetChanged();
    }

    public void clearSelection() {
        clearSelectionInternal();
        notifyDataSetChanged();
    }

    private void clearSelectionInternal() {
        selectionMode = false;
        selectedPositions.clear();
        notifySelectionChanged();
    }

    private void toggleSelection(int position) {
        if (selectedPositions.contains(position)) {
            selectedPositions.remove(position);
        } else {
            selectedPositions.add(position);
        }
        selectionMode = !selectedPositions.isEmpty();
        notifySelectionChanged();
        notifyItemChanged(position);
    }

    private void notifySelectionChanged() {
        if (selectionChangedListener != null) {
            selectionChangedListener.onSelectionChanged(selectedPositions.size(), selectionMode);
        }
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        int layoutId;
        if (viewType == VIEW_TYPE_VIDEO_LIST) {
            layoutId = R.layout.item_video_list;
        } else if (viewType == VIEW_TYPE_VIDEO_TILE) {
            layoutId = R.layout.item_video_tile;
        } else if (viewType == VIEW_TYPE_TILE_DEFAULT) {
            layoutId = R.layout.item_video_tile;
        } else {
            layoutId = R.layout.item_row;
        }
        View view = LayoutInflater.from(parent.getContext()).inflate(layoutId, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        DisplayItem item = items.get(position);
        holder.itemView.setOnClickListener(v -> {
            if (selectionMode) {
                int adapterPosition = holder.getBindingAdapterPosition();
                if (adapterPosition != RecyclerView.NO_POSITION) {
                    toggleSelection(adapterPosition);
                }
                return;
            }
            if (clickListener != null) {
                clickListener.onItemClick(item);
            }
        });
        holder.itemView.setOnLongClickListener(v -> {
            int adapterPosition = holder.getBindingAdapterPosition();
            if (adapterPosition == RecyclerView.NO_POSITION) {
                return false;
            }
            toggleSelection(adapterPosition);
            return true;
        });

        if (item.getType() == DisplayItem.Type.VIDEO) {
            bindVideo(holder, item);
        } else {
            bindDefault(holder, item);
        }

        boolean selected = selectedPositions.contains(position);
        holder.itemView.setBackgroundColor(selected
            ? holder.itemView.getContext().getColor(R.color.selection_bg)
            : Color.TRANSPARENT);
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    @Override
    public int getItemViewType(int position) {
        DisplayItem item = items.get(position);
        if (item.getType() == DisplayItem.Type.VIDEO) {
            return videoDisplayMode == VideoDisplayMode.TILE
                ? VIEW_TYPE_VIDEO_TILE
                : VIEW_TYPE_VIDEO_LIST;
        }
        if (videoDisplayMode == VideoDisplayMode.TILE) {
            return VIEW_TYPE_TILE_DEFAULT;
        }
        return VIEW_TYPE_DEFAULT;
    }

    static class ViewHolder extends RecyclerView.ViewHolder {
        final TextView title;
        final TextView subtitle;
        final ImageView thumbnail;
        final ImageView folderIcon;
        final ImageView overflowButton;
        final int defaultTitleColor;

        ViewHolder(@NonNull View itemView) {
            super(itemView);
            title = itemView.findViewById(R.id.title);
            subtitle = itemView.findViewById(R.id.subtitle);
            thumbnail = itemView.findViewById(R.id.thumbnail);
            folderIcon = itemView.findViewById(R.id.folderIcon);
            overflowButton = itemView.findViewById(R.id.overflowButton);
            defaultTitleColor = title != null ? title.getCurrentTextColor() : Color.BLACK;
        }
    }

    private void bindDefault(ViewHolder holder, DisplayItem item) {
        holder.title.setText(item.getTitle());
        holder.title.setTextColor(holder.defaultTitleColor);
        String subtitle = item.getSubtitle();
        if (subtitle == null || subtitle.isEmpty()) {
            holder.subtitle.setVisibility(View.GONE);
        } else {
            holder.subtitle.setText(subtitle);
            holder.subtitle.setVisibility(View.VISIBLE);
        }

        if (holder.overflowButton != null) {
            holder.overflowButton.setOnClickListener(v -> {
                if (overflowClickListener != null) {
                    overflowClickListener.onOverflowClick(item);
                }
            });
        }

        if (holder.thumbnail != null) {
            holder.thumbnail.setImageResource(android.R.drawable.ic_menu_agenda);
            holder.thumbnail.setScaleType(ImageView.ScaleType.CENTER_INSIDE);
            holder.thumbnail.setColorFilter(holder.thumbnail.getContext().getColor(R.color.brand_green));
        }

        if (holder.folderIcon != null) {
            boolean showIcon = item.getType() == DisplayItem.Type.FOLDER
                || item.getType() == DisplayItem.Type.HIERARCHY;
            holder.folderIcon.setVisibility(showIcon ? View.VISIBLE : View.GONE);
        }

        if (holder.thumbnail != null) {
            int basePadding = dpToPx(holder.itemView, 8);
            holder.itemView.setPadding(basePadding, basePadding, basePadding, basePadding);
        } else {
            int basePadding = dpToPx(holder.itemView, 12);
            int extra = dpToPx(holder.itemView, item.getIndentLevel() * 12);
            holder.itemView.setPadding(basePadding + extra, basePadding, basePadding, basePadding);
        }
    }

    private void bindVideo(ViewHolder holder, DisplayItem item) {
        holder.title.setText(item.getTitle());
        holder.title.setTextColor(holder.defaultTitleColor);
        String details = formatDetails(item);
        holder.subtitle.setText(details);
        holder.subtitle.setVisibility(View.VISIBLE);

        if (holder.overflowButton != null) {
            holder.overflowButton.setOnClickListener(v -> {
                if (overflowClickListener != null) {
                    overflowClickListener.onOverflowClick(item);
                }
            });
        }

        if (holder.thumbnail != null) {
            String uri = item.getContentUri();
            if (uri != null && !uri.isEmpty()) {
                holder.thumbnail.clearColorFilter();
                Glide.with(holder.thumbnail.getContext())
                    .load(Uri.parse(uri))
                    .centerCrop()
                    .into(holder.thumbnail);
            } else {
                holder.thumbnail.setImageDrawable(null);
            }
        }

        int basePadding = dpToPx(holder.itemView, 8);
        holder.itemView.setPadding(basePadding, basePadding, basePadding, basePadding);
    }

    private String formatDetails(DisplayItem item) {
        String duration = formatDuration(item.getDurationMs());
        String resolution = "";
        if (item.getWidth() > 0 && item.getHeight() > 0) {
            resolution = item.getWidth() + "x" + item.getHeight();
        }
        if (!resolution.isEmpty()) {
            return String.format(Locale.US, "%s • %s", duration, resolution);
        }
        return duration;
    }

    private String formatDuration(long durationMs) {
        long totalSeconds = durationMs / 1000;
        long seconds = totalSeconds % 60;
        long minutes = (totalSeconds / 60) % 60;
        long hours = totalSeconds / 3600;
        if (hours > 0) {
            return String.format(Locale.US, "%d:%02d:%02d", hours, minutes, seconds);
        }
        return String.format(Locale.US, "%d:%02d", minutes, seconds);
    }

    private static int dpToPx(View view, int dp) {
        float density = view.getResources().getDisplayMetrics().density;
        return Math.round(dp * density);
    }

}

File: app/src/main/java/com/nshell/nsplayer/ui/VideoListAdapter.kt
package com.nshell.nsplayer.ui

import android.graphics.Color
import android.net.Uri
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.nshell.nsplayer.R
import java.util.Locale

class VideoListAdapter : RecyclerView.Adapter<VideoListAdapter.ViewHolder>() {
    fun interface OnItemClickListener {
        fun onItemClick(item: DisplayItem)
    }

    fun interface OnItemOverflowClickListener {
        fun onOverflowClick(item: DisplayItem)
    }

    fun interface OnSelectionChangedListener {
        fun onSelectionChanged(selectedCount: Int, selectionMode: Boolean)
    }

    private val items = mutableListOf<DisplayItem>()
    private val selectedPositions = mutableSetOf<Int>()
    private var videoDisplayMode = VideoDisplayMode.LIST
    private var clickListener: OnItemClickListener? = null
    private var overflowClickListener: OnItemOverflowClickListener? = null
    private var selectionChangedListener: OnSelectionChangedListener? = null
    private var selectionMode = false

    fun submit(nextItems: List<DisplayItem>?) {
        items.clear()
        if (nextItems != null) {
            items.addAll(nextItems)
        }
        clearSelectionInternal()
        notifyDataSetChanged()
    }

    fun setVideoDisplayMode(mode: VideoDisplayMode?) {
        if (mode != null) {
            videoDisplayMode = mode
        }
    }

    fun setOnItemClickListener(listener: OnItemClickListener?) {
        clickListener = listener
    }

    fun setOnItemOverflowClickListener(listener: OnItemOverflowClickListener?) {
        overflowClickListener = listener
    }

    fun setOnSelectionChangedListener(listener: OnSelectionChangedListener?) {
        selectionChangedListener = listener
    }

    fun isSelectionMode(): Boolean = selectionMode

    fun getSelectedCount(): Int = selectedPositions.size

    fun isAllSelected(): Boolean = items.isNotEmpty() && selectedPositions.size == items.size

    fun getSelectedItems(): List<DisplayItem> {
        if (selectedPositions.isEmpty()) {
            return emptyList()
        }
        return selectedPositions.sorted().mapNotNull { index -> items.getOrNull(index) }
    }

    fun selectAll() {
        selectionMode = true
        selectedPositions.clear()
        for (i in items.indices) {
            selectedPositions.add(i)
        }
        notifySelectionChanged()
        notifyDataSetChanged()
    }

    fun clearSelection() {
        clearSelectionInternal()
        notifyDataSetChanged()
    }

    private fun clearSelectionInternal() {
        selectionMode = false
        selectedPositions.clear()
        notifySelectionChanged()
    }

    private fun toggleSelection(position: Int) {
        if (selectedPositions.contains(position)) {
            selectedPositions.remove(position)
        } else {
            selectedPositions.add(position)
        }
        selectionMode = selectedPositions.isNotEmpty()
        notifySelectionChanged()
        notifyItemChanged(position)
    }

    private fun notifySelectionChanged() {
        selectionChangedListener?.onSelectionChanged(selectedPositions.size, selectionMode)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val layoutId = when (viewType) {
            VIEW_TYPE_VIDEO_LIST -> R.layout.item_video_list
            VIEW_TYPE_VIDEO_TILE -> R.layout.item_video_tile
            VIEW_TYPE_TILE_DEFAULT -> R.layout.item_video_tile
            else -> R.layout.item_row
        }
        val view = LayoutInflater.from(parent.context).inflate(layoutId, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val item = items[position]
        holder.itemView.setOnClickListener {
            if (selectionMode) {
                val adapterPosition = holder.bindingAdapterPosition
                if (adapterPosition != RecyclerView.NO_POSITION) {
                    toggleSelection(adapterPosition)
                }
                return@setOnClickListener
            }
            clickListener?.onItemClick(item)
        }
        holder.itemView.setOnLongClickListener {
            val adapterPosition = holder.bindingAdapterPosition
            if (adapterPosition == RecyclerView.NO_POSITION) {
                return@setOnLongClickListener false
            }
            toggleSelection(adapterPosition)
            true
        }

        if (item.type == DisplayItem.Type.VIDEO) {
            bindVideo(holder, item)
        } else {
            bindDefault(holder, item)
        }

        val selected = selectedPositions.contains(position)
        holder.itemView.setBackgroundColor(
            if (selected) holder.itemView.context.getColor(R.color.selection_bg) else Color.TRANSPARENT
        )
    }

    override fun getItemCount(): Int = items.size

    override fun getItemViewType(position: Int): Int {
        val item = items[position]
        if (item.type == DisplayItem.Type.VIDEO) {
            return if (videoDisplayMode == VideoDisplayMode.TILE) VIEW_TYPE_VIDEO_TILE else VIEW_TYPE_VIDEO_LIST
        }
        return if (videoDisplayMode == VideoDisplayMode.TILE) VIEW_TYPE_TILE_DEFAULT else VIEW_TYPE_DEFAULT
    }

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val title: TextView? = itemView.findViewById(R.id.title)
        val subtitle: TextView? = itemView.findViewById(R.id.subtitle)
        val subtitleRow: View? = itemView.findViewById(R.id.subtitleRow)
        val durationText: TextView? = itemView.findViewById(R.id.durationText)
        val resolutionText: TextView? = itemView.findViewById(R.id.resolutionText)
        val subtitleDivider: View? = itemView.findViewById(R.id.subtitleDivider)
        val thumbnail: ImageView? = itemView.findViewById(R.id.thumbnail)
        val folderIcon: ImageView? = itemView.findViewById(R.id.folderIcon)
        val overflowButton: ImageView? = itemView.findViewById(R.id.overflowButton)
        val defaultTitleColor: Int = title?.currentTextColor ?: Color.BLACK
    }

    private fun bindDefault(holder: ViewHolder, item: DisplayItem) {
        holder.title?.text = item.title
        holder.title?.setTextColor(holder.defaultTitleColor)
        val subtitle = item.subtitle
        if (subtitle.isNullOrEmpty()) {
            holder.subtitle?.visibility = View.GONE
        } else {
            holder.subtitle?.text = subtitle
            holder.subtitle?.visibility = View.VISIBLE
        }
        holder.subtitleRow?.visibility = View.GONE

        holder.overflowButton?.setOnClickListener {
            overflowClickListener?.onOverflowClick(item)
        }

        holder.thumbnail?.let { thumbnail ->
            val isFolder = item.type == DisplayItem.Type.FOLDER || item.type == DisplayItem.Type.HIERARCHY
            val iconRes = if (isFolder) R.drawable.ic_folder else R.drawable.ic_video
            thumbnail.setImageResource(iconRes)
            thumbnail.scaleType = if (isFolder) {
                ImageView.ScaleType.FIT_CENTER
            } else {
                ImageView.ScaleType.CENTER_INSIDE
            }
            thumbnail.setPadding(0, 0, 0, 0)
            thumbnail.setColorFilter(thumbnail.context.getColor(R.color.brand_green))
        }

        holder.folderIcon?.visibility = if (
            item.type == DisplayItem.Type.FOLDER || item.type == DisplayItem.Type.HIERARCHY
        ) {
            View.VISIBLE
        } else {
            View.GONE
        }

        if (holder.thumbnail != null) {
            val basePadding = dpToPx(holder.itemView, 8)
            holder.itemView.setPadding(basePadding, basePadding, basePadding, basePadding)
        } else {
            val basePadding = dpToPx(holder.itemView, 12)
            val extra = dpToPx(holder.itemView, item.indentLevel * 12)
            holder.itemView.setPadding(basePadding + extra, basePadding, basePadding, basePadding)
        }
    }

    private fun bindVideo(holder: ViewHolder, item: DisplayItem) {
        holder.title?.text = item.title
        holder.title?.setTextColor(holder.defaultTitleColor)
        bindSubtitle(holder, item)

        holder.overflowButton?.setOnClickListener {
            overflowClickListener?.onOverflowClick(item)
        }

        holder.thumbnail?.let { thumbnail ->
            val uri = item.contentUri
            if (!uri.isNullOrEmpty()) {
                thumbnail.clearColorFilter()
                Glide.with(thumbnail.context)
                    .load(Uri.parse(uri))
                    .centerCrop()
                    .into(thumbnail)
            } else {
                thumbnail.setImageDrawable(null)
            }
        }

        val basePadding = dpToPx(holder.itemView, 8)
        holder.itemView.setPadding(basePadding, basePadding, basePadding, basePadding)
    }

    private fun bindSubtitle(holder: ViewHolder, item: DisplayItem) {
        val duration = formatDuration(item.durationMs)
        val resolution = if (item.width > 0 && item.height > 0) {
            "${item.width}x${item.height}"
        } else {
            ""
        }
        holder.subtitle?.visibility = View.GONE
        holder.subtitleRow?.visibility = View.VISIBLE
        holder.durationText?.text = duration
        holder.resolutionText?.text = resolution
        val hasResolution = resolution.isNotEmpty()
        holder.subtitleDivider?.visibility = if (hasResolution) View.VISIBLE else View.GONE
        holder.resolutionText?.visibility = if (hasResolution) View.VISIBLE else View.GONE
    }

    private fun formatDuration(durationMs: Long): String {
        val totalSeconds = durationMs / 1000
        val seconds = totalSeconds % 60
        val minutes = (totalSeconds / 60) % 60
        val hours = totalSeconds / 3600
        return if (hours > 0) {
            String.format(Locale.US, "%d:%02d:%02d", hours, minutes, seconds)
        } else {
            String.format(Locale.US, "%d:%02d", minutes, seconds)
        }
    }

    companion object {
        private const val VIEW_TYPE_DEFAULT = 0
        private const val VIEW_TYPE_VIDEO_LIST = 1
        private const val VIEW_TYPE_VIDEO_TILE = 2
        private const val VIEW_TYPE_TILE_DEFAULT = 3

        private fun dpToPx(view: View, dp: Int): Int {
            val density = view.resources.displayMetrics.density
            return Math.round(dp * density)
        }
    }
}
